File: apps/api/internal/scoring/ranking.go
package scoring

import (
	"sort"

	"stockastic/api/internal/money"
	"stockastic/api/internal/rulebook"
)

// Standing is one team's Phase-1 result (Sec 4/5).
type Standing struct {
	AccountID string
	// FinalValue is cash + holdings at the freeze price.
	FinalValue money.Paise
	// PeakValue is the highest portfolio value seen during Phase 1 (tie-break 1).
	PeakValue money.Paise
	// Transactions is the executed-transaction count (tie-break 2: fewer wins).
	Transactions int
}

// DecidedBy names what separated a team from the one ranked immediately above it.
type DecidedBy string

const (
	DecidedByValue      DecidedBy = "value"
	DecidedByPeak       DecidedBy = "peak_portfolio_value"
	DecidedByFewerTxns  DecidedBy = "fewer_transactions"
	DecidedByCoinToss   DecidedBy = "coin_toss"
	DecidedByUnresolved DecidedBy = "unresolved"
)

type Ranked struct {
	Standing
	Rank      int
	DecidedBy DecidedBy
}

// RankPhase1 ranks by final value descending (Sec 4), breaking ties with the configured step
// order (Sec 5). The coin toss is an organiser-supervised manual step, so it is injected as a
// stable per-team key rather than randomised here. A tie that no configured step separates is
// reported as DecidedByUnresolved and ordered by account id so the output is deterministic.
func RankPhase1(standings []Standing, tieBreak []rulebook.TieBreakStep, coinToss func(accountID string) int64) []Ranked {
	if coinToss == nil {
		coinToss = func(string) int64 { return 0 }
	}
	decide := func(a, b Standing) (int, DecidedBy) {
		if a.FinalValue != b.FinalValue {
			if a.FinalValue > b.FinalValue {
				return -1, DecidedByValue
			}
			return 1, DecidedByValue
		}
		for _, step := range tieBreak {
			switch step {
			case rulebook.TieBreakPeakValue:
				if a.PeakValue != b.PeakValue {
					if a.PeakValue > b.PeakValue {
						return -1, DecidedByPeak
					}
					return 1, DecidedByPeak
				}
			case rulebook.TieBreakFewerTransactions:
				if a.Transactions != b.Transactions {
					if a.Transactions < b.Transactions {
						return -1, DecidedByFewerTxns
					}
					return 1, DecidedByFewerTxns
				}
			case rulebook.TieBreakCoinToss:
				ca, cb := coinToss(a.AccountID), coinToss(b.AccountID)
				if ca != cb {
					if ca < cb {
						return -1, DecidedByCoinToss
					}
					return 1, DecidedByCoinToss
				}
			}
		}
		if a.AccountID < b.AccountID {
			return -1, DecidedByUnresolved
		}
		return 1, DecidedByUnresolved
	}

	sorted := append([]Standing(nil), standings...)
	sort.SliceStable(sorted, func(i, j int) bool {
		o, _ := decide(sorted[i], sorted[j])
		return o < 0
	})
	out := make([]Ranked, len(sorted))
	for i, s := range sorted {
		by := DecidedByValue
		if i > 0 {
			_, by = decide(sorted[i-1], s)
		}
		out[i] = Ranked{Standing: s, Rank: i + 1, DecidedBy: by}
	}
	return out
}

File: packages/settlement/src/ranking.ts
import type { TieBreakStep } from "@stockastic/config";

export interface Phase1Standing {
  accountId: string;
  /** cash + holdings at freeze price. */
  finalValue: number;
  /** Highest portfolio value seen during Phase 1 (Sec 5 tie-break 1). */
  peakValue: number;
  /** Executed transactions during Phase 1 (Sec 5 tie-break 2: fewer wins). */
  executedTransactions: number;
}

export type RankDecidedBy = "value" | TieBreakStep | "unresolved";

export interface RankedStanding extends Phase1Standing {
  rank: number;
  /** What separated this team from the one ranked immediately above it. */
  decidedBy: RankDecidedBy;
}

const paise = (v: number) => Math.round(v * 100);

/**
 * Sec 4/5: rank descending by final value; ties broken by the configured
 * step order. `coinToss` must return a stable per-team key (an
 * organiser-supervised toss result) — the rulebook makes this a manual step,
 * so it is injected rather than randomised here.
 */
export function rankPhase1(
  standings: Phase1Standing[],
  tieBreak: TieBreakStep[],
  coinToss: (accountId: string) => number = () => 0
): RankedStanding[] {
  const compareStep = (step: TieBreakStep, a: Phase1Standing, b: Phase1Standing): number => {
    switch (step) {
      case "peak_portfolio_value":
        return paise(b.peakValue) - paise(a.peakValue);
      case "fewer_transactions":
        return a.executedTransactions - b.executedTransactions;
      case "coin_toss":
        return coinToss(a.accountId) - coinToss(b.accountId);
    }
  };

  const decide = (a: Phase1Standing, b: Phase1Standing): { order: number; by: RankDecidedBy } => {
    const byValue = paise(b.finalValue) - paise(a.finalValue);
    if (byValue !== 0) return { order: byValue, by: "value" };
    for (const step of tieBreak) {
      const c = compareStep(step, a, b);
      if (c !== 0) return { order: c, by: step };
    }
    return { order: a.accountId < b.accountId ? -1 : 1, by: "unresolved" };
  };

  const sorted = [...standings].sort((a, b) => decide(a, b).order);
  return sorted.map((s, i) => ({
    ...s,
    rank: i + 1,
    decidedBy: i === 0 ? "value" : decide(sorted[i - 1], s).by,
  }));
}
